apps/www/_blog/2022-12-14-mfa-auth-via-rls.mdx

```diff
@@ -0,0 +1,149 @@
+
+**Enrollment**
+
+To generate a QR code, call the `/enroll` endpoint which returns an SVG encoded QR and the secret. Thereafter, create a challenge by calling the `/challenge` endpoint. Once the user has entered the six digit TOTP code generated by their authenticator app, call the`/verify`  endpoint with the corresponding factor and challenge details.
+
+You might wonder: why the need for the challenge step? The challenge step is designed to create an interval between MFA initiation and the action of making a verification. This is useful in cases like Yubikey authentication where a user might need to request a challenge before placing their finger on the device.
+
+```mermaid
+graph TD
+    Z[Start] --> A
+    A[Enroll Factor] --> B(Challenge Factor) -->  K(Verify Factor)
+    K -->|Correct Code| D[Factor Verified, Logged in ]
+    K -->|Incorrect Code/Error| Z
+
+```
+
+**Verification**
+
+On subsequent logins attempts, redirect a user to an MFA verification after they have completed the conventional sign in process. On the verification page, wait for the user to enter the six digit OTP code from the authenticator application and then call  the`/challenge` endpoint followed by the `/verify` endpoint. If a correct code is submitted, a JWT will be created with a few additional fields.
+
+```mermaid
+graph TD
+    Z[Start] --> K(Challenge Factor)
+    K ---> L(Verify Factor)
+    L -->|Correct Code| D[Factor Verified, Logged in ]
+    L -->|Incorrect Code/Error| Z
+
+```
+
+## Enforcement via Row Level Security
+
+Adding MFA to your project and not enforcing it is like buying an expensive door and never locking it. We love Postgres RLS at Supabase and we believe it plays a key role in facilitating custom checks for a spread of use cases. To support RLS integration, JWTs issued by Supabase Auth now contain two additional pieces of information:
+
+1. An Authenticator Assurance Level (AAL) claim. Use this to quickly identify the level of identity checks the user has performed. `aal1` is reserved for conventional sign in, while `aal2` is issued only after the user has verified with an additional factor.
+
+2. An Authenticator Method Reference (AMR) claim. Use this to identify all of the authentication methods used by the user. This is also useful if you wish to implement step-up login scenarios.
+
+```jsx
+{
+  "aud": "authenticated",
+  "exp": 1670929371,
+  "sub": "8802c1d6-c555-46e3-aacd-b61198b058d9",
+  "email": "j0@supacats.io",
+  "phone": "",
+  "app_metadata": {
+    "provider": "email",
+    "providers": [
+      "email"
+    ]
+  },
+  "user_metadata": {},
+  "role": "authenticated",
+  "aal": "aal2",
+  "amr": [
+    {
+      "method": "password",
+      "timestamp": 1670924394
+    },
+    {
+      "method": "totp",
+      "timestamp": 1670925771
+    }
+  ],
+  "session_id": "9db34b97-a0e3-4eaf-810c-00b21b844cd0"
+}
+```
```

Comment:
May be better to just highlight the important bits?

```suggestion
```jsx
{
  "sub": "8802c1d6-c555-46e3-aacd-b61198b058d9",
  "email": "j0@supacats.io",
  "aud": "authenticated",
  "exp": 1670929371,
  "aal": "aal2",
  "amr": [
    {
      "method": "password",
      "timestamp": 1670924394
    },
    {
      "method": "totp",
      "timestamp": 1670925771
    }
  ],
  // ...
}
```
```